File: tools/hollow-knight-patches/src/skins/core/SkinRotationCore.cs
using System.Linq;

namespace DualSouls.Skins.HollowKnight.Core
{
    // Task64 selector only. Task65 owns mode/proof/transaction composition and correlated retirement.
    // INTENT_ISSUED is deliberately parked: no new death, second delivery or outcome authority here.
    public sealed class SkinRotationCore
    {
        // H1 emits mode-only CAS values. Task66 delegates visual execution to TransactionCore.
        public RotationModeDecision Decide(RotationModeState s, RotationModeEvent signal)
        {
            if (!IsModeStateValid(s)) return Mode(s,"invalid-state");
            if (signal == null) return Mode(s,"invalid-event");
            if (signal is RotationModeEvent.Selector selector)
            {
                if (!ValidEvent(selector.Event)) return Mode(s,"invalid-event");
                if (selector.Event is RotationEvent.Rebind)
                {
                    var next=Decide(s.Rotation,selector.Event).State;
                    if(next==s.Rotation)return Mode(s,"same-binding");
                    return Mode(s with {Rotation=next,LiveProof=null,Operation=s.Operation == null ? null : s.Operation with {ReboundBlocked=true}},"rebound");
                }
                if(s.OffRequested || s.Operation!=null)return Mode(s,"mode-busy");
                var result=Decide(s.Rotation,selector.Event);
                return Mode(s with {Rotation=result.State},result.Diagnosis);
            }
            if(signal is RotationModeEvent.VerifiedVisual verified)
            {
                if(!ProofValid(verified.Value) || verified.Value.Hero!=s.Rotation.CurrentHero || verified.Value.Proof.Binding!=s.Rotation.CurrentSkin)return Mode(s,"stale-proof");
                if(s.Operation!=null)return Mode(s,"mode-busy");
                return Mode(s with {LiveProof=verified.Value},"visual-verified");
            }
            if(signal is RotationModeEvent.AdvanceMode)return Advance(s);
            var p=s.Operation;if(p==null)return Mode(s,"no-mode-operation");
            var correlation=signal switch {
                RotationModeEvent.ModeCommitted e=>e.Correlation,
                RotationModeEvent.ModeIndeterminate e=>e.Correlation,
                RotationModeEvent.ModeFailed e=>e.Correlation,
                RotationModeEvent.RetryModeCommit e=>e.Correlation,
                _=>null
            };
            if(correlation==null || correlation!=p.Correlation)return Mode(s,"stale-correlation");
            if(p.ReboundBlocked)return Mode(s,"rebound-resolution-required");
            switch(signal)
            {
                case RotationModeEvent.ModeCommitted complete:
                    var r=complete.Receipt;var h=complete.Head;
                    if(r==null || !Uuid(r.NewGenerationId) || !RotationValues.Digest(r.NewGenerationSha256) ||
                        r.ExpectedGenerationId!=p.BaseHead.GenerationId || r.ExpectedGenerationSha256!=p.BaseHead.GenerationSha256 ||
                        r.NewGenerationId==r.ExpectedGenerationId || r.NewGenerationSha256==r.ExpectedGenerationSha256 ||
                        !HeadValid(h) || h.GenerationId!=r.NewGenerationId || h.GenerationSha256!=r.NewGenerationSha256 || h.Activation!=p.Target)
                        return Mode(s,"stale-receipt");
                    return Mode(s with {Rotation=s.Rotation with {Activation=p.Target},Head=h,Operation=null,OffRequested=false,CanceledPending=null},"mode-committed");
                case RotationModeEvent.ModeIndeterminate _:
                    return Mode(s with {Operation=p with {Phase=ModeCommitPhase.INDETERMINATE}},"mode-indeterminate");
                case RotationModeEvent.ModeFailed _:
                    return Mode(s with {Operation=p with {Phase=ModeCommitPhase.DEFINITIVE_FAILURE}},"mode-failed");
                case RotationModeEvent.RetryModeCommit _:
                    return Mode(s,"retry-exact-cas",Command(p));
                default:return Mode(s,"invalid-event");
            }
        }
        private static RotationModeDecision Advance(RotationModeState initial)
        {
            if(initial.Operation!=null)return Mode(initial,"mode-busy");
            if(initial.Rotation.CurrentHero==null)return Mode(initial,"unbound");
            var a=initial.Rotation.Activation;
            if(a.Mode==SkinMode.OFF)
            {
                if(a.SelectedPackId==null)return Mode(initial,"NO_SELECTED_SKIN");
                var selected=initial.Rotation.Ring.Entries.SingleOrDefault(x=>x.CurrentObject.Id==a.SelectedPackId);
                if(selected==null)return Mode(initial,"selected-object-unavailable");
                if(!SelectedEstablished(initial.LiveProof,a.Active,selected.CurrentObject))return Mode(initial,"transaction-required");
                if(initial.OperationHighWater==long.MaxValue)return Mode(initial,"operation-id-exhausted");
                var onId=initial.OperationHighWater+1;
                // No visual write: preserve stamp, but publish the exact CURRENT import identity.
                var on=new ModeOperation(new(OperationId(onId),initial.Rotation.CurrentHero,initial.Rotation.CurrentSkin),
                    initial.Head,a with {Mode=SkinMode.ON,Active=selected.CurrentObject},SelectedProof:initial.LiveProof);
                return Mode(initial with {OperationHighWater=onId,Operation=on},"commit-mode",Command(on));
            }
            var s=initial;
            if(a.Mode==SkinMode.ROTATE)
            {
                var pending=s.Rotation.Pending;s=s with {OffRequested=true};
                if(pending?.Phase==RotationPendingPhase.INTENT_ISSUED)return Mode(s,"issued-rotation-busy");
                if(pending!=null)s=s with {Rotation=s.Rotation with {Pending=null},CanceledPending=pending};
                if(!(s.LiveProof?.Proof is VerifiedLiveVisualProof.Vanilla))return Mode(s,"transaction-required");
            }
            if(s.OperationHighWater==long.MaxValue)return Mode(s,"operation-id-exhausted");
            var next=s.OperationHighWater+1;
            var p=new ModeOperation(new(OperationId(next),s.Rotation.CurrentHero,s.Rotation.CurrentSkin),s.Head,
                a with {Mode=a.Mode==SkinMode.ON ? SkinMode.ROTATE : SkinMode.OFF},a.Mode==SkinMode.ROTATE ? s.LiveProof : null);
            return Mode(s with {OperationHighWater=next,Operation=p},"commit-mode",Command(p));
        }
        // Validates the entire composed state; this is NOT an Apply permission query.
        public bool IsModeStateValid(RotationModeState s)
        {
            if(s==null || !Valid(s.Rotation) || !HeadValid(s.Head) || s.Head.Activation!=s.Rotation.Activation || s.OperationHighWater<0)return false;
            if(s.LiveProof!=null && (!ProofValid(s.LiveProof) || s.LiveProof.Hero!=s.Rotation.CurrentHero || s.LiveProof.Proof.Binding!=s.Rotation.CurrentSkin))return false;
            if(s.OffRequested && s.Rotation.Activation.Mode!=SkinMode.ROTATE)return false;
            var canceled=s.CanceledPending;
            if(canceled!=null && (!s.OffRequested || s.Rotation.Pending!=null || canceled.Phase!=RotationPendingPhase.AWAITING_STABILITY ||
                !Valid(s.Rotation with {CurrentHero=canceled.Hero,CurrentSkin=canceled.Skin,Pending=canceled})))return false;
            if(s.OffRequested && s.Rotation.Pending?.Phase==RotationPendingPhase.AWAITING_STABILITY)return false;
            var p=s.Operation;if(p==null)return true;
            if(s.Rotation.CurrentHero==null || s.Rotation.Pending!=null || s.OperationHighWater==0 || p.Correlation==null ||
                p.Correlation.OperationId!=OperationId(s.OperationHighWater) || !Hero(p.Correlation.Hero) || !Skin(p.Correlation.Skin) ||
                p.BaseHead!=s.Head || !RotationValues.Activation(p.Target) ||
                (p.Phase!=ModeCommitPhase.ISSUED && p.Phase!=ModeCommitPhase.INDETERMINATE && p.Phase!=ModeCommitPhase.DEFINITIVE_FAILURE))return false;
            if(!p.ReboundBlocked && (p.Correlation.Hero!=s.Rotation.CurrentHero || p.Correlation.Skin!=s.Rotation.CurrentSkin))return false;
            if(p.ReboundBlocked && s.LiveProof!=null)return false;
            if(!p.ReboundBlocked && p.VanillaProof!=null && s.LiveProof!=p.VanillaProof)return false;
            if(!p.ReboundBlocked && p.SelectedProof!=null && s.LiveProof!=p.SelectedProof)return false;
            var a=s.Rotation.Activation;
            if(a.Mode!=SkinMode.OFF && p.SelectedProof!=null)return false;
            var selected=s.Rotation.Ring.Entries.SingleOrDefault(x=>x.CurrentObject.Id==a.SelectedPackId);
            return a.Mode switch {
                SkinMode.OFF => selected!=null && !s.OffRequested && p.VanillaProof==null &&
                    p.Target==(a with {Mode=SkinMode.ON,Active=selected.CurrentObject}) &&
                    SelectedEstablished(p.SelectedProof,a.Active,selected.CurrentObject) &&
                    p.SelectedProof.Hero==p.Correlation.Hero && p.SelectedProof.Proof.Binding==p.Correlation.Skin,
                SkinMode.ON => !s.OffRequested && p.VanillaProof==null && p.Target==(a with {Mode=SkinMode.ROTATE}),
                SkinMode.ROTATE => s.OffRequested && p.Target==(a with {Mode=SkinMode.OFF}) && ProofValid(p.VanillaProof) &&
                    p.VanillaProof.Proof is VerifiedLiveVisualProof.Vanilla && p.VanillaProof.Hero==p.Correlation.Hero && p.VanillaProof.Proof.Binding==p.Correlation.Skin,
                _=>false
            };
        }
        public bool IsPriorEstablished(RotationModeState s)
        {
            if(!IsModeStateValid(s))return false;
            var proof=s.LiveProof?.Proof;var active=s.Rotation.Activation.Active;
            return proof is VerifiedLiveVisualProof.Vanilla && active is ActiveVisual.Vanilla ||
                proof is VerifiedLiveVisualProof.Pack p && active is ActiveVisual.Pack a &&
                p.Visual.Id==a.Id && p.Visual.TreeSha256==a.TreeSha256 && p.Visual.ContentSha256==a.ContentSha256;
        }
        private static bool SelectedEstablished(HeroVerifiedVisual p,ActiveVisual prior,ActiveVisual.Pack target)=>
            ProofValid(p) && p.Proof is VerifiedLiveVisualProof.Pack proof && prior is ActiveVisual.Pack a &&
            SamePackVisual(proof.Visual,a) && SamePackVisual(proof.Visual,target);
        private static bool SamePackVisual(ActiveVisual.Pack a,ActiveVisual.Pack b)=>
            a.Id==b.Id && a.TreeSha256==b.TreeSha256 && a.ContentSha256==b.ContentSha256;
        private static bool ProofValid(HeroVerifiedVisual p)=>p!=null && Hero(p.Hero) && p.Proof!=null && Skin(p.Proof.Binding) &&
            (p.Proof is VerifiedLiveVisualProof.Vanilla || p.Proof is VerifiedLiveVisualProof.Pack v && RotationValues.Pack(v.Visual));
        private static bool HeadValid(VerifiedRegistryHead h)=>h!=null && Uuid(h.GenerationId) && RotationValues.Digest(h.GenerationSha256) &&
            RotationValues.Activation(h.Activation) && h.Interlock==RotationInterlock.Clear();
        private static bool Uuid(string s)=>s!=null && System.Text.RegularExpressions.Regex.IsMatch(s,"\\A[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\\z");
        private static string OperationId(long n)
        {
            var hex=n.ToString("x16",System.Globalization.CultureInfo.InvariantCulture);
            return "00000000-0000-"+hex.Substring(0,4)+"-"+hex.Substring(4,4)+"-0000"+hex.Substring(8);
        }
        private static RotationModeCommand Command(ModeOperation p)
        {
            var issued=p with {Phase=ModeCommitPhase.ISSUED};
            return p.Target.Mode==SkinMode.ON ? new RotationModeCommand.CommitVerifiedSelectedOn(issued) :
                p.Target.Mode==SkinMode.ROTATE ? new RotationModeCommand.CommitOnToRotate(issued) : new RotationModeCommand.CommitVerifiedVanillaOff(issued);
        }
        private static RotationModeDecision Mode(RotationModeState s,string diagnosis,params RotationModeCommand[] commands)=>new(s,diagnosis,commands);

        public RotationDecision Decide(RotationState state, RotationEvent signal)
        {
            if (!Valid(state)) return D(state,"invalid-state");
            if (!ValidEvent(signal)) return D(state,"invalid-event");
            if (signal is RotationEvent.Rebind binding)
            {
                if(binding.Hero==state.CurrentHero && binding.Skin==state.CurrentSkin) return D(state,"same-binding");
                return D(state with {CurrentHero=binding.Hero,CurrentSkin=binding.Skin,
                    Pending=state.Pending == null ? null : state.Pending with {Hero=binding.Hero,Skin=binding.Skin}},"rebound");
            }
            if(state.CurrentHero==null) return D(state,"unbound");
            return signal switch
            {
                RotationEvent.ConfirmDeath e => Confirm(state,e),
                RotationEvent.StableRespawn e => Stable(state,e.Token),
                _ => D(state,"invalid-event")
            };
        }
        private static RotationDecision Confirm(RotationState s, RotationEvent.ConfirmDeath e)
        {
            if(e.Hero!=s.CurrentHero || e.Skin!=s.CurrentSkin) return D(s,"stale-binding");
            if(s.Activation.Mode!=SkinMode.ROTATE) return D(s,"rotation-inactive");
            if(e.Epoch.Value==0 || e.Epoch.Value<=s.EpochHighWater.Value) return D(s,"consumed-epoch");
            if(s.Pending!=null) return D(s,"pending-rotation");
            var candidate=Successor(s);if(candidate==null) return D(s,"insufficient-eligible");
            return D(s with {EpochHighWater=e.Epoch,Pending=new PendingRotation(candidate,e.Epoch,e.Hero,e.Skin)},"candidate-confirmed");
        }
        private static RotationDecision Stable(RotationState s,StableRespawnToken token)
        {
            if(token.Hero!=s.CurrentHero || token.Skin!=s.CurrentSkin) return D(s,"stale-binding");
            var p=s.Pending;if(p==null) return D(s,"no-pending");
            if(token.DeathEpoch!=p.Epoch) return D(s,"stale-epoch");
            if(p.Phase==RotationPendingPhase.INTENT_ISSUED) return D(s,"intent-already-issued");
            var intent=new RotationReadyIntent(p.Epoch,p.Hero,p.Skin,p.Candidate,s.Activation);
            return new RotationDecision(s with {Pending=p with {Phase=RotationPendingPhase.INTENT_ISSUED,IssuedIntent=intent}},"rotation-ready",intent);
        }
        private static RotationDescriptor Successor(RotationState s)
        {
            var eligible=s.Ring.Entries.Where(x=>x.Eligible).ToList();if(eligible.Count<=1)return null;
            var active=(s.Activation.Active as ActiveVisual.Pack)?.Id;
            var anchor=eligible.FindIndex(x=>x.CurrentObject.Id==active);
            if(anchor<0)anchor=eligible.FindIndex(x=>x.CurrentObject.Id==s.Activation.SelectedPackId);
            return eligible[(anchor+1)%eligible.Count];
        }
        private static bool Valid(RotationState s)
        {
            if(s==null || s.Ring==null || s.EpochHighWater==null || !RotationValues.Activation(s.Activation) ||
                (s.CurrentHero==null)!=(s.CurrentSkin==null))return false;
            if(s.CurrentHero==null)return s.Pending==null && s.EpochHighWater.Value==0;
            if(!Hero(s.CurrentHero)||!Skin(s.CurrentSkin))return false;
            var p=s.Pending;if(p==null)return true;
            if(s.Activation.Mode!=SkinMode.ROTATE || p.Epoch==null || p.Epoch.Value==0 || p.Epoch!=s.EpochHighWater ||
                p.Hero!=s.CurrentHero || p.Skin!=s.CurrentSkin || p.Candidate==null || p.Candidate!=Successor(s))return false;
            return p.Phase switch
            {
                RotationPendingPhase.AWAITING_STABILITY => p.IssuedIntent==null,
                RotationPendingPhase.INTENT_ISSUED => p.IssuedIntent!=null && p.IssuedIntent.Epoch==p.Epoch &&
                    p.IssuedIntent.Candidate==p.Candidate && p.IssuedIntent.Prior==s.Activation &&
                    Hero(p.IssuedIntent.Hero) && Skin(p.IssuedIntent.Skin),
                _ => false
            };
        }
        private static bool Hero(HeroBindingToken h)=>h!=null && RotationValues.Text(h.Value,256);
        private static bool Skin(SkinBindingToken s)=>s!=null && RotationValues.Text(s.Value,256);
        private static bool ValidEvent(RotationEvent e)=>e switch
        {
            RotationEvent.Rebind b=>Hero(b.Hero)&&Skin(b.Skin),
            RotationEvent.ConfirmDeath d=>d.Epoch!=null&&Hero(d.Hero)&&Skin(d.Skin),
            RotationEvent.StableRespawn r=>r.Token!=null&&r.Token.DeathEpoch!=null&&Hero(r.Token.Hero)&&Skin(r.Token.Skin),
            _=>false
        };
        private static RotationDecision D(RotationState s,string diagnosis)=>new(s,diagnosis);
    }
}

File: src/SilksongLauncher.Launcher/app/src/main/kotlin/dev/silksong/launcher/skins/core/SkinRotationCore.kt
package dev.silksong.launcher.skins.core

import dev.silksong.launcher.skins.registry.*

/** Task64 selector only. Task65 owns mode/proof/transaction composition and correlated retirement.
 * INTENT_ISSUED is deliberately parked: no new death, second delivery or outcome authority here. */
class SkinRotationCore {
    /** H1 emits mode-only CAS values; task66 must delegate all visual execution to TransactionCore. */
    fun decide(s: RotationModeState, event: RotationModeEvent): RotationModeDecision {
        if (!isModeStateValid(s)) return mode(s, "invalid-state")
        if (event is RotationModeEvent.Selector) {
            if (!validEvent(event.event)) return mode(s, "invalid-event")
            if (event.event is RotationEvent.Rebind) {
                val next = decide(s.rotation, event.event).state
                if (next == s.rotation) return mode(s, "same-binding")
                return mode(s.copy(rotation = next, liveProof = null,
                    operation = s.operation?.copy(reboundBlocked = true)), "rebound")
            }
            if (s.offRequested || s.operation != null) return mode(s, "mode-busy")
            val next = decide(s.rotation, event.event)
            return mode(s.copy(rotation = next.state), next.diagnosis)
        }
        if (event is RotationModeEvent.VerifiedVisual) {
            if (!proofValid(event.value) || event.value.hero != s.rotation.currentHero ||
                event.value.proof.binding != s.rotation.currentSkin) return mode(s, "stale-proof")
            if (s.operation != null) return mode(s, "mode-busy")
            return mode(s.copy(liveProof = event.value), "visual-verified")
        }
        if (event is RotationModeEvent.AdvanceMode) return advance(s)
        val p = s.operation ?: return mode(s, "no-mode-operation")
        val correlation = when (event) {
            is RotationModeEvent.ModeCommitted -> event.correlation
            is RotationModeEvent.ModeIndeterminate -> event.correlation
            is RotationModeEvent.ModeFailed -> event.correlation
            is RotationModeEvent.RetryModeCommit -> event.correlation
            else -> return mode(s, "invalid-event")
        }
        if (correlation != p.correlation) return mode(s, "stale-correlation")
        if (p.reboundBlocked) return mode(s, "rebound-resolution-required")
        return when (event) {
            is RotationModeEvent.ModeCommitted -> {
                val r = event.receipt; val h = event.head
                if (!uuid(r.newGenerationId) || !digest(r.newGenerationSha256) ||
                    r.expectedGenerationId != p.baseHead.generationId || r.expectedGenerationSha256 != p.baseHead.generationSha256 ||
                    r.newGenerationId == r.expectedGenerationId || r.newGenerationSha256 == r.expectedGenerationSha256 ||
                    !headValid(h) || h.generationId != r.newGenerationId || h.generationSha256 != r.newGenerationSha256 ||
                    h.activation != p.target) mode(s, "stale-receipt")
                else mode(s.copy(rotation = s.rotation.copy(activation = p.target), head = h,
                    operation = null, offRequested = false, canceledPending = null), "mode-committed")
            }
            is RotationModeEvent.ModeIndeterminate -> mode(s.copy(operation = p.copy(phase = ModeCommitPhase.INDETERMINATE)), "mode-indeterminate")
            is RotationModeEvent.ModeFailed -> mode(s.copy(operation = p.copy(phase = ModeCommitPhase.DEFINITIVE_FAILURE)), "mode-failed")
            is RotationModeEvent.RetryModeCommit -> mode(s, "retry-exact-cas", command(p))
            else -> mode(s, "invalid-event")
        }
    }
    private fun advance(initial: RotationModeState): RotationModeDecision {
        if (initial.operation != null) return mode(initial, "mode-busy")
        if (initial.rotation.currentHero == null) return mode(initial, "unbound")
        val a = initial.rotation.activation
        if (a.mode == SkinMode.OFF) {
            if (a.selectedPackId == null) return mode(initial, "NO_SELECTED_SKIN")
            val selected = initial.rotation.ring.entries.singleOrNull { it.currentObject.id == a.selectedPackId }
                ?: return mode(initial, "selected-object-unavailable")
            if (!selectedEstablished(initial.liveProof, a.active, selected.currentObject)) return mode(initial, "transaction-required")
            if (initial.operationHighWater == Long.MAX_VALUE) return mode(initial, "operation-id-exhausted")
            val next = initial.operationHighWater + 1
            // Visual proof ignores import identity; the exact CAS publishes CURRENT identity.
            // No visual write means no new skin stamp, including at stamp exhaustion.
            val p = ModeOperation(ModeCorrelation(operationId(next), initial.rotation.currentHero, initial.rotation.currentSkin!!),
                initial.head, a.copy(mode = SkinMode.ON, active = selected.currentObject), selectedProof = initial.liveProof)
            return mode(initial.copy(operationHighWater = next, operation = p), "commit-mode", command(p))
        }
        var s = initial
        if (a.mode == SkinMode.ROTATE) {
            val pending = s.rotation.pending
            s = s.copy(offRequested = true)
            if (pending?.phase == RotationPendingPhase.INTENT_ISSUED) return mode(s, "issued-rotation-busy")
            if (pending != null) s = s.copy(rotation = s.rotation.copy(pending = null), canceledPending = pending)
            if (s.liveProof?.proof !is VerifiedLiveVisualProof.Vanilla) return mode(s, "transaction-required")
        }
        if (s.operationHighWater == Long.MAX_VALUE) return mode(s, "operation-id-exhausted")
        val next = s.operationHighWater + 1
        val p = ModeOperation(ModeCorrelation(operationId(next), s.rotation.currentHero!!, s.rotation.currentSkin!!),
            s.head, a.copy(mode = if (a.mode == SkinMode.ON) SkinMode.ROTATE else SkinMode.OFF),
            if (a.mode == SkinMode.ROTATE) s.liveProof else null)
        return mode(s.copy(operationHighWater = next, operation = p), "commit-mode", command(p))
    }
    /** Validates the entire composed state; this is NOT an Apply permission query. */
    fun isModeStateValid(s: RotationModeState): Boolean {
        if (!valid(s.rotation) || !headValid(s.head) || s.head.activation != s.rotation.activation || s.operationHighWater < 0) return false
        if (s.liveProof != null && (!proofValid(s.liveProof) || s.liveProof.hero != s.rotation.currentHero ||
                s.liveProof.proof.binding != s.rotation.currentSkin)) return false
        if (s.offRequested && s.rotation.activation.mode != SkinMode.ROTATE) return false
        val canceled = s.canceledPending
        if (canceled != null) {
            if (!s.offRequested || s.rotation.pending != null || canceled.phase != RotationPendingPhase.AWAITING_STABILITY ||
                !valid(s.rotation.copy(currentHero = canceled.hero, currentSkin = canceled.skin, pending = canceled))) return false
        }
        if (s.offRequested && s.rotation.pending?.phase == RotationPendingPhase.AWAITING_STABILITY) return false
        val p = s.operation ?: return true
        if (s.rotation.currentHero == null || s.rotation.pending != null || s.operationHighWater == 0L ||
            p.correlation.operationId != operationId(s.operationHighWater) || !RotationValues.text(p.correlation.hero.value, 256) ||
            !RotationValues.text(p.correlation.skin.value, 256) || p.baseHead != s.head || !RotationValues.activation(p.target)) return false
        if (!p.reboundBlocked && (p.correlation.hero != s.rotation.currentHero || p.correlation.skin != s.rotation.currentSkin)) return false
        if (p.reboundBlocked && s.liveProof != null) return false
        if (!p.reboundBlocked && p.vanillaProof != null && s.liveProof != p.vanillaProof) return false
        if (!p.reboundBlocked && p.selectedProof != null && s.liveProof != p.selectedProof) return false
        val a = s.rotation.activation
        if (a.mode != SkinMode.OFF && p.selectedProof != null) return false
        return when (a.mode) {
            SkinMode.ON -> !s.offRequested && p.vanillaProof == null && p.target == a.copy(mode = SkinMode.ROTATE)
            SkinMode.ROTATE -> s.offRequested && p.target == a.copy(mode = SkinMode.OFF) && p.vanillaProof?.let {
                proofValid(it) && it.proof is VerifiedLiveVisualProof.Vanilla && it.hero == p.correlation.hero && it.proof.binding == p.correlation.skin
            } == true
            SkinMode.OFF -> s.rotation.ring.entries.singleOrNull { it.currentObject.id == a.selectedPackId }?.let {
                !s.offRequested && p.vanillaProof == null && p.target == a.copy(mode = SkinMode.ON, active = it.currentObject) &&
                    p.selectedProof?.hero == p.correlation.hero && p.selectedProof.proof.binding == p.correlation.skin &&
                    selectedEstablished(p.selectedProof, a.active, it.currentObject)
            } == true
        }
    }
    fun isPriorEstablished(s: RotationModeState): Boolean {
        if (!isModeStateValid(s)) return false
        val proof = s.liveProof?.proof ?: return false
        val active = s.rotation.activation.active
        return proof is VerifiedLiveVisualProof.Vanilla && active is ActiveVisual.Vanilla ||
            proof is VerifiedLiveVisualProof.Pack && active is ActiveVisual.Pack && proof.visual.id == active.id &&
            proof.visual.treeSha256 == active.treeSha256 && proof.visual.contentSha256 == active.contentSha256
    }
    private fun selectedEstablished(p: HeroVerifiedVisual?, prior: ActiveVisual, target: ActiveVisual.Pack): Boolean =
        p != null && proofValid(p) && p.proof is VerifiedLiveVisualProof.Pack && prior is ActiveVisual.Pack &&
            samePackVisual(p.proof.visual, prior) && samePackVisual(p.proof.visual, target)
    private fun samePackVisual(a: ActiveVisual.Pack, b: ActiveVisual.Pack) =
        a.id == b.id && a.treeSha256 == b.treeSha256 && a.contentSha256 == b.contentSha256
    private fun proofValid(p: HeroVerifiedVisual) = RotationValues.text(p.hero.value, 256) &&
        RotationValues.text(p.proof.binding.value, 256) && when (val v = p.proof) {
            is VerifiedLiveVisualProof.Vanilla -> true
            is VerifiedLiveVisualProof.Pack -> RotationValues.pack(v.visual)
        }
    private fun headValid(h: VerifiedRegistryHead) = uuid(h.generationId) && digest(h.generationSha256) &&
        RotationValues.activation(h.activation) && h.interlock == RotationInterlock.clear()
    private fun uuid(s: String) = Regex("[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}").matches(s)
    private fun digest(s: String) = Regex("[0-9a-f]{64}").matches(s)
    private fun operationId(n: Long): String {
        val hex = n.toString(16).padStart(16, '0')
        return "00000000-0000-${hex.substring(0, 4)}-${hex.substring(4, 8)}-0000${hex.substring(8)}"
    }
    private fun command(p: ModeOperation): RotationModeCommand {
        val issued = p.copy(phase = ModeCommitPhase.ISSUED)
        return if (p.target.mode == SkinMode.ON) RotationModeCommand.CommitVerifiedSelectedOn(issued)
        else if (p.target.mode == SkinMode.ROTATE) RotationModeCommand.CommitOnToRotate(issued)
        else RotationModeCommand.CommitVerifiedVanillaOff(issued)
    }
    private fun mode(s: RotationModeState, diagnosis: String, vararg commands: RotationModeCommand) = RotationModeDecision(s, diagnosis, *commands)

    fun decide(state: RotationState, event: RotationEvent): RotationDecision {
        if (!valid(state)) return d(state, "invalid-state")
        if (!validEvent(event)) return d(state, "invalid-event")
        if (event is RotationEvent.Rebind) {
            if (event.hero == state.currentHero && event.skin == state.currentSkin) return d(state, "same-binding")
            return d(state.copy(currentHero = event.hero, currentSkin = event.skin,
                pending = state.pending?.copy(hero = event.hero, skin = event.skin)), "rebound")
        }
        if (state.currentHero == null) return d(state, "unbound")
        return when (event) {
            is RotationEvent.ConfirmDeath -> confirm(state, event)
            is RotationEvent.StableRespawn -> stable(state, event.token)
            is RotationEvent.Rebind -> d(state, "invalid-event")
        }
    }
    private fun confirm(s: RotationState, e: RotationEvent.ConfirmDeath): RotationDecision {
        if (e.hero != s.currentHero || e.skin != s.currentSkin) return d(s, "stale-binding")
        if (s.activation.mode != SkinMode.ROTATE) return d(s, "rotation-inactive")
        if (e.epoch.value == 0uL || e.epoch.value <= s.epochHighWater.value) return d(s, "consumed-epoch")
        if (s.pending != null) return d(s, "pending-rotation")
        val candidate = successor(s) ?: return d(s, "insufficient-eligible")
        return d(s.copy(epochHighWater = e.epoch, pending = PendingRotation(candidate, e.epoch, e.hero, e.skin)), "candidate-confirmed")
    }
    private fun stable(s: RotationState, token: StableRespawnToken): RotationDecision {
        if (token.hero != s.currentHero || token.skin != s.currentSkin) return d(s, "stale-binding")
        val p = s.pending ?: return d(s, "no-pending")
        if (token.deathEpoch != p.epoch) return d(s, "stale-epoch")
        if (p.phase == RotationPendingPhase.INTENT_ISSUED) return d(s, "intent-already-issued")
        val intent = RotationReadyIntent(p.epoch, p.hero, p.skin, p.candidate, s.activation)
        return RotationDecision(s.copy(pending = p.copy(phase = RotationPendingPhase.INTENT_ISSUED, issuedIntent = intent)), "rotation-ready", intent)
    }
    private fun successor(s: RotationState): RotationDescriptor? {
        val eligible = s.ring.entries.filter { it.eligible }
        if (eligible.size <= 1) return null
        val active = (s.activation.active as? ActiveVisual.Pack)?.id
        var anchor = eligible.indexOfFirst { it.currentObject.id == active }
        if (anchor < 0) anchor = eligible.indexOfFirst { it.currentObject.id == s.activation.selectedPackId }
        return eligible[(anchor + 1) % eligible.size]
    }
    private fun valid(s: RotationState): Boolean {
        if (!RotationValues.activation(s.activation) || (s.currentHero == null) != (s.currentSkin == null)) return false
        if (s.currentHero == null) return s.pending == null && s.epochHighWater.value == 0uL
        if (!RotationValues.text(s.currentHero.value, 256) || !RotationValues.text(s.currentSkin!!.value, 256)) return false
        val p = s.pending ?: return true
        if (s.activation.mode != SkinMode.ROTATE || p.epoch.value == 0uL || p.epoch != s.epochHighWater ||
            p.hero != s.currentHero || p.skin != s.currentSkin || p.candidate != successor(s)) return false
        return when (p.phase) {
            RotationPendingPhase.AWAITING_STABILITY -> p.issuedIntent == null
            RotationPendingPhase.INTENT_ISSUED -> p.issuedIntent?.let {
                it.epoch == p.epoch && it.candidate == p.candidate && it.prior == s.activation &&
                    RotationValues.text(it.hero.value, 256) && RotationValues.text(it.skin.value, 256)
            } == true
        }
    }
    private fun validEvent(e: RotationEvent): Boolean = when (e) {
        is RotationEvent.Rebind -> RotationValues.text(e.hero.value, 256) && RotationValues.text(e.skin.value, 256)
        is RotationEvent.ConfirmDeath -> RotationValues.text(e.hero.value, 256) && RotationValues.text(e.skin.value, 256)
        is RotationEvent.StableRespawn -> RotationValues.text(e.token.hero.value, 256) && RotationValues.text(e.token.skin.value, 256)
    }
    private fun d(s: RotationState, diagnosis: String) = RotationDecision(s, diagnosis)
}
